### PR TITLE
Fixed algorithm

### PR DESCRIPTION
**Pull request by esuwu**

### CODE REVIEW

File: pkg/storing/events/events.go [hunk 1]
```diff
@@ -174,9 +191,51 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
+			continue
+		}
+		nodesHeights = append(nodesHeights, entities.NodeStatement{Node: node, Height: minHeight, Status: entities.OK})
+	}
+
+	return nodesHeights, nodesList
+}
+
+func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.NodeStatement, error) {
+	var heightDifference = depthCommonHeightSearch
```
Comment (nickeskov) [inline pkg/storing/events/events.go:214]: `heightDifference` variable is redundant.

File: pkg/storing/events/events.go [hunk 2]
```diff
@@ -174,9 +191,51 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
+	nodesList := make(map[string]bool) // reachable or unreachable
+	for _, node := range nodes {
+		nodesList[node] = true
+	}
+
+	nodesHeights, nodesList, minHeight := s.FindMinCommonLatestHeight(nodesList, minHeight, maxHeight)
+
 	if (maxHeight - minHeight) > heightDifference {
 		return nodesHeights, BigHeightDifference
 	}
```
Comment (nickeskov) [inline pkg/storing/events/events.go:227]: `maxHeight - minHeight < 0` in any case. This check is redundant. I think it should be placed below next `for` cycle.

File: pkg/storing/events/events.go [hunk 3]
```diff
@@ -174,9 +191,51 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
+	var heightDifference = depthCommonHeightSearch
+
+	minHeight, maxHeight := math.MaxInt, 0
+
+	nodesList := make(map[string]bool) // reachable or unreachable
+	for _, node := range nodes {
+		nodesList[node] = true
+	}
+
+	nodesHeights, nodesList, minHeight := s.FindMinCommonLatestHeight(nodesList, minHeight, maxHeight)
```
Comment (nickeskov) [inline pkg/storing/events/events.go:223]: `maxHeight == 0` in any case. Is it necessary here? I think nope.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/storing/events/events.go [hunk 4]
```diff
@@ -174,9 +191,51 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
+
+func (s *Storage) FindMinCommonSpecificHeight(nodesList map[string]bool, minHeight int) ([]entities.NodeStatement, map[string]bool) {
+	// looking for the min common height and the max height
+	var nodesHeights []entities.NodeStatement
+	for node := range nodesList {
+		found, err := s.FoundStatementAtHeight(node, minHeight)
+		if err != nil || !found {
+			nodesList[node] = false // this node is unreachable
+			nodesHeights = append(nodesHeights, entities.NodeStatement{Node: node, Status: entities.Unreachable})
+			continue
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/storing/events/events.go:205]: It is not really an error. The case of a node being unreachable is handled and it's a routine case. Even if it was an error, the behavior would not be any different

File: pkg/storing/events/events.go (hunk 2 above)
Comment (esuwu) [reply pkg/storing/events/events.go:227]: Where do you see `0`?

File: pkg/storing/events/events.go (hunk 3 above)
Comment (esuwu) [reply pkg/storing/events/events.go:223]: My bad, the variable is not supposed to be copied to FindMinCommonLatestHeight. It is supposed to be changed there

File: pkg/storing/events/events.go [hunk 5]
```diff
@@ -174,9 +191,51 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
 		return nodesHeights, BigHeightDifference
 	}
+
+	for i := 0; i < depthCommonHeightSearch; i++ {
+		for _, node := range nodesHeights {
+			if node.Height != minHeight {
+				minHeight = minHeight - 1
+				nodesHeights, nodesList = s.FindMinCommonSpecificHeight(nodesList, minHeight)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/storing/events/events.go:235]: The comment looks strange. Yes, FindMinCommonSpecificHeight changes nodesHeights just as it's supposed to. What's the question?

File: pkg/storing/events/events.go (hunk 2 above)
Comment (nickeskov) [reply pkg/storing/events/events.go:227]: Above. (`minHeight, maxHeight := math.MaxInt, 0`)

Comment (nickeskov) [reply pkg/storing/events/events.go:227]: I mean `maxHeight - minHeight` is negative, so the check result is always `true`.

File: pkg/storing/events/events.go [hunk 6]
```diff
@@ -142,19 +146,32 @@ func (s *Storage) LatestHeight(node string) (int, error) {
 	return h, nil
 }
 
+func (s *Storage) FoundStatementAtHeight(node string, height int) (bool, error) {
+	pattern := newStatementKey(node, "*")
+	var st entities.NodeStatement
+	err := s.viewByKeyPatternWithDescendKeys(pattern, func(s *entities.NodeStatement) bool {
+		st := *s
+		if st.Height != height {
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/storing/events/events.go:154]: If the statement is not found for a specific node, we start looking on `height-1` again and again for every node until we find all statements on a specific height. If not found afterwards, the monitoring is going to say that storage does not contain enough statements

[the remark this answers is not in the document]
Comment (nickeskov) [reply pkg/storing/events/events.go:154]: Okay, understood. Agree.

File: pkg/storing/events/events.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nickeskov) [reply pkg/storing/events/events.go:235]: It's not a question. It's a recommendation to refactor this. Changing range's source variable inside cycle body isn't obvious It's quite hard to read such code.

[the remark this answers is not in the document]
Comment (nickeskov) [reply pkg/storing/events/events.go:235]: For instance, run the test:
```go
func TestChangeSliceWhileIterating(t *testing.T) {
	src := []byte{3, 2, 1, 0}
	var res []byte
	for _, v := range src {
		if v == 0 {
			src = []byte{10}
		}
		res = append(res, v)
	}
	if !bytes.Equal(src, res) {
		t.Error("bytes are not equal")
	}
}
```

File: pkg/storing/events/events.go (hunk 1 above)
Comment (esuwu) [reply pkg/storing/events/events.go:214]: It's not in this case. `heightDifference` just happened to be the same as `depthCommonHeightSearch` (10), so I decided to make one constant

**Comment by esuwu** [comment]
@nickeskov Fixed, check again pls

File: pkg/storing/events/events.go [hunk 7]
```diff
@@ -137,28 +145,79 @@ func (s *Storage) LatestHeight(node string) (int, error) {
...
+	notFound := false
+	err := s.viewByKeyPatternWithDescendKeys(pattern, func(s *entities.NodeStatement) bool {
+		st = *s
+		if st.Height < height {
+			notFound = true
+			return false
+		}
+		if st.Height != height {
+			return true
+		} else {
```
Comment (nickeskov) [inline pkg/storing/events/events.go:165]: Unnecessary `else`.

File: pkg/storing/events/events.go [hunk 8]
```diff
@@ -137,28 +145,79 @@ func (s *Storage) LatestHeight(node string) (int, error) {
...
-	var heightDifference = 10
-
-	minHeight, maxHeight := math.MaxInt, 0
+func (s *Storage) StatementAtHeight(node string, height int) (entities.NodeStatement, error) {
+	pattern := newStatementKey(node, "*")
+	var st entities.NodeStatement
+	notFound := false
+	err := s.viewByKeyPatternWithDescendKeys(pattern, func(s *entities.NodeStatement) bool {
+		st = *s
+		if st.Height < height {
```
Comment (nickeskov) [inline pkg/storing/events/events.go:159]: We also need to handle rollback cases. For example, change the `if` condition like this:
```go
const maxHeightThreshold = 100
if st.Height < height && height - st.Height > maxHeightThreshold {
    // snip
}
```

File: pkg/storing/events/events.go [hunk 9]
```diff
@@ -137,28 +145,79 @@ func (s *Storage) LatestHeight(node string) (int, error) {
...
+		return false, errors.Wrapf(err, "failed to get the last state hash at height %d for node '%s'", height, node)
 	}
+	if notFound {
+		return false, nil
+	}
+	if st.StateHash == nil {
+		return false, errors.Wrapf(NoFullStatementError, "no full statement at height %d for node '%s'", height, node)
+	}
+	return true, nil
+}
```
Comment (nickeskov) [inline pkg/storing/events/events.go:207]: This is almost the same as `StatementAtHeight` method.

File: pkg/storing/events/events.go [hunk 10]
```diff
@@ -174,40 +233,88 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
-			}
-			return true
-		})
+		statement, err := s.StatementAtHeight(node, minHeight)
 		if err != nil {
-			return nil, err
+			if !errors.Is(err, NoFullStatementError) {
+				return nil, errors.Wrapf(err, "failed to find statement at height %d for node '%s'", minHeight, node)
+			}
+			fmt.Printf("failed to find a statement for node %s on height %d: %v\n", node, minHeight, err)
```
Comment (nickeskov) [inline pkg/storing/events/events.go:308]: Use `log.Printf`.

File: pkg/storing/events/events.go [hunk 11]
```diff
@@ -174,40 +233,88 @@ func (s *Storage) FindAllStatehashesOnCommonHeight(nodes []string) ([]entities.N
...
+			continue
 		}
-
-		if !foundStatementOnHeight {
-			fmt.Printf("NOT FOUND STATEMENT FOR NODE %s\n", node)
-			return nil, StorageIsNotReady
+		if statement.Height == minHeight {
+			statementsOnHeight = append(statementsOnHeight, statement)
+		} else {
+			fmt.Printf("wrong height in statement for node %s on min height %d\n, received %d\n", node, minHeight, statement.Height)
```
Comment (nickeskov) [inline pkg/storing/events/events.go:315]: Use log.Printf.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/storing/events/events.go (hunk 8 above)
Comment (esuwu) [reply pkg/storing/events/events.go:159]: This function is not supposed to do that. It looks for a value at an exact height. There are other checks for rollback cases

**Review by nickeskov — APPROVED** [review]

**Pull request merged**